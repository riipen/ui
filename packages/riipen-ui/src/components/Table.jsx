import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";
import withForwardedRef from "../utils/withForwardedRef";

class Table extends React.Component {
  static propTypes = {
    /**
     * What background to render for the table.
     * Defaults to grey100.
     */
    backgroundColor: PropTypes.oneOf([
      "transparent",
      "grey50",
      "grey100",
      "grey200",
      "grey300",
      "grey400"
    ]),

    /**
     * Children to render in the table element.
     */
    children: PropTypes.node,

    /**
     * Classes to apply to the root element.
     */
    classes: PropTypes.arrayOf(PropTypes.string),

    /**
     * A reference to the table DOM element.
     */
    forwardedRef: PropTypes.element,

    /**
     * Whether to give all columns in the table equal width.
     */
    layout: PropTypes.oneOf(["auto", "fixed"])
  };

  static defaultProps = {
    backgroundColor: "grey100",
    layout: "auto"
  };

  static contextType = ThemeContext;

  render() {
    const {
      children,
      classes,
      layout,
      backgroundColor,
      forwardedRef
    } = this.props;

    const theme = this.context;

    const tableBackground = {
      transparent: "transparent",
      grey50: theme.palette.grey[50],
      grey100: theme.palette.grey[100],
      grey200: theme.palette.grey[200],
      grey300: theme.palette.grey[300],
      grey400: theme.palette.grey[400]
    };

    return (
      <React.Fragment>
        <table ref={forwardedRef} className={clsx(classes)} border="0">
          {children}
        </table>
        <style jsx>{`
          table {
            background-color: ${tableBackground[backgroundColor]};
            border-radius: 4px;
            border-collapse: collapse;
            table-layout: ${layout};
            width: 100%;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withForwardedRef(withClasses(Table));
