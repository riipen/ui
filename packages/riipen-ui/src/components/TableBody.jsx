import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import withClasses from "../utils/withClasses";

class TableBody extends React.Component {
  static propTypes = {
    /**
     * Children to render in the tbody element
     */
    children: PropTypes.node,

    /**
     * Classes to apply to the root element
     */
    classes: PropTypes.arrayOf(PropTypes.string)
  };

  render() {
    const { children, classes, ...other } = this.props;
    return (
      <tbody classes={clsx(classes)} {...other}>
        {children}
      </tbody>
    );
  }
}

export default withClasses(TableBody);
