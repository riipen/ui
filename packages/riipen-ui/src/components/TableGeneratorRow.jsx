import PropTypes from "prop-types";
import React, { useState } from "react";
import _JSXStyle from "styled-jsx/style";

import ThemeContext from "../styles/ThemeContext";

import withClasses from "../utils/withClasses";

import Table from "./Table";
import TableRow from "./TableRow";
import TableBody from "./TableBody";
import TableCell from "./TableCell";
import TableHeaderCell from "./TableHeaderCell";

const TableGeneratorRow = ({
  hover,
  data,
  columns,
  expandedNode,
  expandRow,
  isLastRow,
  mobile,
  rowNum,
  rowProps
}) => {
  const [hovering, setHovering] = useState(false);
  const theme = React.useContext(ThemeContext);

  const handleRowMouseEnter = hover ? () => setHovering(true) : undefined;
  const handleRowMouseLeave = hover ? () => setHovering(false) : undefined;

  const isExpandable = expandRow && expandedNode;
  const isExpanded = isExpandable && expandRow(data);

  const renderDefault = () =>
    columns.map((col, j) => (
      <TableCell
        key={`default-cell-${rowNum}-${j}`}
        {...col.cellProps?.(data, rowNum, mobile)}
      >
        {col?.cell(data, rowNum, mobile)}
      </TableCell>
    ));

  const renderMobile = () => {
    const mobileHeader = columns.find(x => x.mobileHeader);
    const mobileFooter = columns.find(x => x.mobileFooter);
    const bodyColumns = columns.filter(x => !x.mobileFooter && !x.mobileHeader);

    return (
      <TableCell padding={0}>
        {mobileHeader && (
          <Table backgroundColor="transparent">
            <TableBody>
              <TableRow>
                <TableCell>
                  {mobileHeader?.cell(data, rowNum, mobile)}
                </TableCell>
              </TableRow>
            </TableBody>
          </Table>
        )}
        <Table layout="fixed" backgroundColor="transparent">
          <TableBody>
            {bodyColumns.map((bodyColumn, j) => (
              <TableRow border={false} key={`${rowNum}-${j}`}>
                <TableHeaderCell align="left">
                  {bodyColumn?.header(mobile)}
                </TableHeaderCell>
                <TableCell>{bodyColumn?.cell(data, rowNum, mobile)}</TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
        {mobileFooter && (
          <Table backgroundColor="transparent">
            <TableBody>
              <TableRow border={false}>
                <TableCell>
                  {mobileFooter?.cell(data, rowNum, mobile)}
                </TableCell>
              </TableRow>
            </TableBody>
          </Table>
        )}
      </TableCell>
    );
  };

  return (
    <>
      <TableRow
        border={!isLastRow || isExpanded}
        forceHover={hovering}
        onMouseEnter={handleRowMouseEnter}
        onMouseLeave={handleRowMouseLeave}
        {...rowProps}
      >
        {!mobile && renderDefault()}
        {mobile && renderMobile()}
      </TableRow>
      {isExpanded && (
        <TableRow
          border={!isLastRow}
          forceHover={hovering}
          onMouseEnter={handleRowMouseEnter}
          onMouseLeave={handleRowMouseLeave}
        >
          <TableCell
            classes={["expanded"]}
            padding={0}
            colSpan={columns.length}
          >
            {expandedNode(data)}
          </TableCell>
        </TableRow>
      )}
      <style jsx>{`
        .expanded {
          background-color: ${theme.palette.grey[100]};
        }
      `}</style>
    </>
  );
};

TableGeneratorRow.displayName = "TableGeneratorRow";

TableGeneratorRow.propTypes = {
  /**
   * Classes to apply to the root element.
   */
  classes: PropTypes.arrayOf(PropTypes.string),

  /**
   * The columns to render.
   */
  columns: PropTypes.arrayOf(
    PropTypes.shape({
      cell: PropTypes.func.isRequired,
      cellProps: PropTypes.func,
      header: PropTypes.func.isRequired,
      headerProps: PropTypes.func,
      mobileFooter: PropTypes.bool,
      mobileHeader: PropTypes.bool
    })
  ),

  /**
   * Row data to render.
   */
  data: PropTypes.object,

  /**
   * Function to determine if row should be expanded or not
   */
  expandRow: PropTypes.func,

  /**
   * Node to render for an expanded row.
   */
  expandedNode: PropTypes.func,

  /**
   * Whether or not to highlight rows on highlight.
   */
  hover: PropTypes.bool,

  /**
   * Is this the last row or not.
   */
  isLastRow: PropTypes.bool,

  /**
   * Render mobile view or not.
   */
  mobile: PropTypes.bool,

  /**
   * The number of this row
   */
  rowNum: PropTypes.number,

  /**
   * Properties to pass to a rendered row.
   */
  rowProps: PropTypes.object
};

TableGeneratorRow.defaultProps = {
  columns: [],
  data: {},
  hover: true,
  mobile: false,
  isLastRow: false
};

export default withClasses(TableGeneratorRow);
