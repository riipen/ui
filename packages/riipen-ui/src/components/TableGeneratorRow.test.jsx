import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Table from "./Table";
import TableBody from "./TableBody";
import TableGeneratorRow from "./TableGeneratorRow";

describe("<TableGeneratorRow>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(
        <Table>
          <TableBody>
            <TableGeneratorRow />
          </TableBody>
        </Table>
      );
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(
      <Table>
        <TableBody>
          <TableGeneratorRow />
        </TableBody>
      </Table>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
