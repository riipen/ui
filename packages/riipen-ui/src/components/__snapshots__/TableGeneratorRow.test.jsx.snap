// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableGeneratorRow> renders correct snapshot 1`] = `
<ForwardRef(withForwardedRef(withClasses(Table)))>
  <withClasses(Table)
    forwardedRef={null}
  >
    <Table
      backgroundColor="grey50"
      classes={
        Array [
          "riipen",
          "riipen-table",
        ]
      }
      forwardedRef={null}
      layout="auto"
    >
      <table
        border="0"
        className="jsx-1059775881 riipen riipen-table"
      >
        <withClasses(TableBody)>
          <TableBody
            classes={
              Array [
                "riipen",
                "riipen-tablebody",
              ]
            }
          >
            <tbody
              classes="riipen riipen-tablebody"
            >
              <withClasses(TableGeneratorRow)>
                <TableGeneratorRow
                  classes={
                    Array [
                      "riipen",
                      "riipen-tablegeneratorrow",
                    ]
                  }
                  columns={Array []}
                  data={Object {}}
                  hover={true}
                  isLastRow={false}
                  mobile={false}
                >
                  <withClasses(TableRow)
                    border={true}
                    forceHover={false}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                  >
                    <TableRow
                      border={true}
                      classes={
                        Array [
                          "riipen",
                          "riipen-tablerow",
                        ]
                      }
                      forceHover={false}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                    >
                      <tr
                        className="jsx-4111142428 riipen riipen-tablerow"
                        onMouseEnter={[Function]}
                        onMouseLeave={[Function]}
                      />
                      <JSXStyle
                        dynamic={
                          Array [
                            "#ddd",
                            "#fff",
                            "#f58632",
                            "",
                            "",
                          ]
                        }
                        id="2808869914"
                      />
                    </TableRow>
                  </withClasses(TableRow)>
                  <JSXStyle
                    dynamic={
                      Array [
                        "#f5f5f5",
                      ]
                    }
                    id="725096434"
                  />
                </TableGeneratorRow>
              </withClasses(TableGeneratorRow)>
            </tbody>
          </TableBody>
        </withClasses(TableBody)>
      </table>
      <JSXStyle
        dynamic={
          Array [
            "#fafafa",
            "auto",
          ]
        }
        id="4123435624"
      />
    </Table>
  </withClasses(Table)>
</ForwardRef(withForwardedRef(withClasses(Table)))>
`;
